text/trace/0107-sampling-score.md

```diff
@@ -0,0 +1,217 @@
+# Associating sampling score with the trace
+
+Enable consistent sampling across distributed application with different
+sampling rates and probability calculation algorithms.
+
+## TL;DR
+
+**Score** is a floating point number associated with
+the trace. It's calculated when trace starts and flows in the `tracestate`,
+it's used by samplers to make consistent sampling decisions.
+
+Service that starts the trace calculates the score and adds it to the
+`tracestate` so downstream services can re-use it to make their sampling
+decisions *instead of* re-calculating score as a function of trace-id
+(or trace-flags).
+
+*Score* is not related to sampling *rate* (aka *probability* which represents
+sampler's configuration not specific to trace).
+
+## Motivation
+
+The goal is to enable a mechanism for consistent (best effort) sampling
+between services with different sampling rates and different probability
+calculation algorithms (for interoperability with existing tracing tools).
+
+Consistent sampling decision made in each app of a distributed trace is
+important for better user experience of trace analysis. Consistency is achieved
+by following means:
+
+1. Same hashing algorithms used across all apps in a trace.
+   Coordination of sampling algorithms across multiple apps not always possible:
+   for example existing components in a system use vendor-specific
+   tracing tool (pre-OpenTelemetry and update is hard to justify) while there
+   is a desire to use OpenTelemetry for new components.
+
+2. Sampling flag propagated from the head component/app is used by downstream
+   apps to sample in a given trace.
+   It requires to trust upstream decision and does not allow to have different
+   sampling rates across different components.
+
+## Explanation
+
+Sampling propabaility is generated by the first service to make sampling
+decision. It's a random float number in [0, 1] range.
+Score is stamped on the span and also propagated further within `tracestate`.
+
+Next service reads score from `tracestate` (instead of calculating it from
+trace-id) and compares it with its sampling rate to make sampling decision.
+
+Score is also exposed through span attributes. Vendors can leverage it
+to sort traces based on their completeness: the lower the value of score is,
+the higher the chance it was sampled it by each component.
```

Comment:
Do we always need to have a score? How do we calculate the score if a custom sampler is used like always on? How does the score interact with custom samplers that do rate based sampling 1 trace every 1 second? How does the backend know when score was used to do sampling decision or just ignored?

Reply:
We don't always need to have a score. We need it when there are multiple possible sampling algorithms in the same app. 
Example when we need it:
```
Service A uses Azure Monitor SDK, uses hash1(trace-id), assigns score  -> Service B uses OTel with hash2(trace-id)
                                                                       -> Service C uses AzureMontior with hash1(trace-id)
```
                                                 
Services A and C use Azure Monitor SDK for years and it's unreasonable to ask them to upgrade to OTel. 
It is reasonable to ask them to update to the new version of Azure Monitor that supports score.

We don't need score when only OTel is used everywhere (or the same algorithm). 

So, we'll tell our customers to configure `ExternalScoreSampler` and fallback to OTel TraceIdRationBased algorithm if score is not in the tracestate.

Default case for OTel users does not change, this is opt-in behavior. 
Custom samplers that use  1 trace every 1 second are not compatible with score - they don't pursuit consistent sampling goal anyway and applying them together with `ExternalScoreSampler` seems to be a misconfiguration.